### PR TITLE
Adds an IDC simulation module to the Debug Helper.

### PR DESCRIPTION
**Pull request by zinigor**
This change adds a new module to the Debug Helper module: 
Screenshot from 2021-09-24 22-52-49

So far the module is pretty simple - it lets you avoid adding filters, while simply enabling a simulated identity crisis using the `jetpack_sync_site_url` filter.

#### Changes proposed in this Pull Request:
* Added an IDC simulation module to the Debug Helper.

#### Jetpack product discussion
* N/A

#### Does this pull request change what data or activity we track or use?
* No

#### Testing instructions:
* Build the `jetpack-debug` or enable it using Jetpack Beta.
* Enable the IDC simulator module in the menu.
* Enable the IDC simulation in the UI.
* Create a draft post,
* Observe your site in IDC safe mode.

### CODE REVIEW

File: projects/plugins/debug-helper/modules/class-idc-simulator.php [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+	public function __construct() {
+		add_action( 'admin_menu', array( $this, 'idc_simulator_register_submenu_page' ), 1000 );
+
+		add_action( 'admin_enqueue_scripts', array( $this, 'enqueue_scripts' ) );
+
+		// Stored options.
+		add_action( 'admin_post_store_current_options', array( $this, 'admin_post_store_current_options' ) );
+
+		$settings                     = $this->get_stored_settings();
+		$this->idc_simulation_enabled = 1 === intval( $settings['idc_simulation'] ) ?
```
Comment (kbrown9) [inline projects/plugins/debug-helper/modules/class-idc-simulator.php:54]: If the option is empty, `get_stored_settings()` will return false, and `$settings['idc_simulation']` will generate a PHP notice:

> PHP Notice:  Trying to access array offset on value of type bool in .../plugins/debug-helper/modules/class-idc-simulator.php on line 54

Should we check whether `$settings['idc_simulation']` exists before trying to use it?

File: projects/plugins/debug-helper/modules/class-idc-simulator.php [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+		new IDC_Simulator();
+	} else {
+		add_action( 'admin_notices', 'idc_simulator_jetpack_not_active' );
+	}
+}
+
+/**
+ * Notice for if Jetpack is not active.
+ */
+function idc_siulator_jetpack_not_active() {
```
Comment (kbrown9) [inline projects/plugins/debug-helper/modules/class-idc-simulator.php:234]: Just a typo:

```suggestion
function idc_simulator_jetpack_not_active() {
```

File: projects/plugins/debug-helper/modules/class-idc-simulator.php [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+			);
+		} else {
+			wp_safe_redirect( get_home_url() );
+		}
+	}
+
+	/**
+	 * Displays an admin notice...
+	 */
+	public function render_admin_notice() {
```
Comment (kbrown9) [inline projects/plugins/debug-helper/modules/class-idc-simulator.php:210]: Is `render_admin_notice()` used? I see a different notice ("_Success! Backup of the connection options stored safely._") when I save the IDC simulation settings.

Comment (zinigor) [reply projects/plugins/debug-helper/modules/class-idc-simulator.php]: You're right, thank you! This was copypasta from the other module - I have removed it in favor of a simplified notice.

File: projects/plugins/debug-helper/modules/class-idc-simulator.php (hunk 2 above)
Comment (zinigor) [reply projects/plugins/debug-helper/modules/class-idc-simulator.php]: Ouch, I didn't see you submitted a suggestion - I have fixed it in another commit, but thanks anyway!

File: projects/plugins/debug-helper/modules/class-idc-simulator.php (hunk 1 above)
Comment (zinigor) [reply projects/plugins/debug-helper/modules/class-idc-simulator.php:54]: I have added `wp_parse_args` to add sane defaults, thanks for the catch!

**Comment by kbrown9** [comment]
I'm running into an issue during testing. The IDC that the simulator generates can be unexpectedly cleared, and it appears to happen because the `IDC_Simulator::spoof_siteurl` callback isn't hooked before [Identity_Crisis::validate_sync_error_idc_option](https://github.com/Automattic/jetpack/blob/master/projects/packages/identity-crisis/src/class-identity-crisis.php#L324) is called. The IDC error isn't validated (because the urls are match) so the `migrate_for_idc` option is set to true, and wpcom changes the siteurl to the spoofed url.

You can reproduce this with the following steps:
1. Enable the IDC Simulator.
2. Do something that causes a sync (edit a post, etc.)
3. Notice that the IDC UI is displayed on wp-admin pages.
4. Do some other things on the site, edit posts, go to wp-admin pages, etc. Eventually, the IDC UI will no longer be displayed. If you check the site's siteurl on wpcom, you'll see that it's changed to the spoof url. I've seen this happen on a request to `wp-admin/admin-ajax.php`.

I'm not sure if this can be fixed by changing the priority of the `register_idc_simulator` hook.

**Review by kbrown9 — APPROVED** [review]
This works well in my tests!

**Pull request merged**